enterprise/backend/src/metabase_enterprise/semantic_search/task/indexer.clj

```diff
@@ -0,0 +1,66 @@
+(ns metabase-enterprise.semantic-search.task.indexer
+  (:require [clojurewerkz.quartzite.jobs :as jobs]
+            [clojurewerkz.quartzite.schedule.simple :as simple]
+            [clojurewerkz.quartzite.triggers :as triggers]
+            [metabase-enterprise.semantic-search.env :as semantic.env]
+            [metabase-enterprise.semantic-search.indexer :as semantic-search.indexer]
+            [metabase.task.core :as task]
+            [metabase.util.log :as log])
+  (:import (java.time Duration Instant)
+           (java.util Date)
+           (org.quartz DisallowConcurrentExecution)))
+
+(set! *warn-on-reflection* true)
+
+(def ^:private indexer-stem
+  (jobs/key "metabase-enterprise.semantic-search.indexer"))
+
+(def indexer-job-key
+  "Key used to define and trigger a job that maintains semantic search indexes."
+  (jobs/key (str indexer-stem ".job")))
+
+;; would prefer a job member, as quartz suggests in its InterruptableJob docs
+;; but if I do that quartz cannot initialize the job - there is probably a way around this.
+(defonce ^:private execution-thread-ref (volatile! nil))
+
+(deftype ^{DisallowConcurrentExecution true
+           :doc                        "Runs an indexer process for a time, expects to be rescheduled to continue"}
+ SemanticSearchIndexer []
+  org.quartz.Job
+  (execute [_ _]
+    (log/with-context {:quartz-job-type 'SemanticSearchIndexer}
+      (try
+        (vreset! execution-thread-ref (Thread/currentThread))
+        (when (semantic.env/pgvector-configured?)
+          (semantic-search.indexer/quartz-job-run! (semantic.env/get-pgvector-datasource!) (semantic.env/get-index-metadata)))
+        (finally
+          (locking execution-thread-ref
+            (vreset! execution-thread-ref nil))))))
+  org.quartz.InterruptableJob
+  (interrupt [_]
+   ;; locking required here to avoid racing with the unset in the finally
+   ;; and interrupting some other unintended task/work
+    (locking execution-thread-ref
+      (when-some [^Thread execution-thread @execution-thread-ref]
+        (.interrupt execution-thread)))))
+
+(def ^:private ^Duration startup-delay (Duration/parse "PT10S"))
+(def ^:private ^Duration run-frequency (Duration/parse "PT10S"))
```

Comment:
This only represents a indexing gap when the gate indexer is cold. Individual runs will take up to an hour given the current configuration - indexing is eventually yielded to quartz - so jobs do not run indefinitely. 

Also note that the cold early exit is defined in the indexer loop, so that if we do not see anything new in some time, we also yield back to quartz. 

I did not want to have an infinite quartz job, intuitively it felt that would cause havoc. Feel free to call me out on this - maybe that its fine!

FWIW Normally with this kind of thing you would do a bit of leader election for each index or something like that: ultimately the goal is ensuring multiple nodes do not clobber the same index.

Reply:
Is there any benefit to having the startup-delay be randomized at all? I feel like I've seen slack threads in the past were cloud restarts a bunch of instances at the same time and they all kickoff some startup job at ~ the same time and smash the appdb (or pgvector db in this case). I guess reindexing on startup takes long enough that a small amount of randomization is unlikely to help much.

Reply:
> I did not want to have an infinite quartz job

1+ Something we could consider for the future, but I think having it yield back is a good place to start

Reply:
> is there any benefit to having the startup-delay be randomized at all

I am not sure this is needed in this case, there should only be one instance of the job running at a time?

Reply:
No no it is a good point! I am not sure how the virtualization of pgvector is going to work, its magic to me. I probably need to learn about that before I can say anything useful!